File: Sources/RTLSSync/SyncEngine.swift
import Foundation
import RTLSCore

public enum SyncEngineEvent: Sendable, Equatable {
    case didUpload(accepted: Int, rejected: Int)
    case uploadFailed(String)
    case didPull(count: Int)
    case pullFailed(String)
}

public actor SyncEngine {
    private let store: LocationStore
    private let bidirectionalStore: (any BidirectionalLocationStore)?
    private let api: LocationSyncAPI
    private let pullAPI: (any LocationPullAPI)?
    private let mergeStrategy: (any LocationMergeStrategy)?
    private let batching: BatchingPolicy
    private let retryPolicy: SyncRetryPolicy
    private let retentionPolicy: RetentionPolicy
    private let network: NetworkMonitor
    private let pullInterval: TimeInterval?
    private let pullOnForeground: Bool

    private var timerTask: Task<Void, Never>?
    private var networkTask: Task<Void, Never>?
    private var pullTask: Task<Void, Never>?
    private var notifyDebounceTask: Task<Void, Never>?
    private var flushing = false

    private var failureCount = 0
    private var nextAllowedFlushAt: Date?
    private var lastPruneAt: Date?
    private var lastPullAt: Date?
    private var pendingInsertsSinceLastFlush = 0

    private let eventsStream: AsyncStream<SyncEngineEvent>
    private let continuation: AsyncStream<SyncEngineEvent>.Continuation
    public nonisolated var events: AsyncStream<SyncEngineEvent> { eventsStream }

    public init(
        store: LocationStore,
        api: LocationSyncAPI,
        batchingPolicy: BatchingPolicy,
        retryPolicy: SyncRetryPolicy = .default,
        retentionPolicy: RetentionPolicy = .recommended,
        network: NetworkMonitor = NetworkMonitor(),
        pullAPI: (any LocationPullAPI)? = nil,
        mergeStrategy: (any LocationMergeStrategy)? = nil,
        pullInterval: TimeInterval? = nil,
        pullOnForeground: Bool = true
    ) {
        self.store = store
        self.bidirectionalStore = store as? any BidirectionalLocationStore
        self.api = api
        self.pullAPI = pullAPI
        self.mergeStrategy = mergeStrategy
        self.batching = batchingPolicy
        self.retryPolicy = retryPolicy
        self.retentionPolicy = retentionPolicy
        self.network = network
        self.pullInterval = pullInterval
        self.pullOnForeground = pullOnForeground

        let (stream, continuation) = AsyncStream<SyncEngineEvent>.makeStream(bufferingPolicy: .bufferingNewest(64))
        self.eventsStream = stream
        self.continuation = continuation
    }

    public func start() {
        Task { await network.start() }

        timerTask?.cancel()
        timerTask = Task { [batching] in
            while !Task.isCancelled {
                let interval = max(1.0, batching.flushInterval)
                try? await Task.sleep(nanoseconds: UInt64(interval * 1_000_000_000))
                await self.flushIfNeeded(force: false)
            }
        }

        networkTask?.cancel()
        networkTask = Task {
            for await online in network.updates {
                if online {
                    await self.flushIfNeeded(force: false)
                }
            }
        }

        if pullAPI != nil, let bidir = bidirectionalStore {
            let interval = pullInterval ?? 60.0
            pullTask?.cancel()
            pullTask = Task { [pullAPI, mergeStrategy] in
                while !Task.isCancelled {
                    try? await Task.sleep(nanoseconds: UInt64(max(1.0, interval) * 1_000_000_000))
                    await self.runPullOnce(pullAPI: pullAPI!, store: bidir, mergeStrategy: mergeStrategy)
                }
            }
        }
    }

    public func stop() {
        timerTask?.cancel()
        timerTask = nil

        networkTask?.cancel()
        networkTask = nil

        pullTask?.cancel()
        pullTask = nil

        Task { await network.stop() }
    }

    public func notifyNewData() {
        pendingInsertsSinceLastFlush += 1
        if pendingInsertsSinceLastFlush >= batching.maxBatchSize {
            Task { await flushIfNeeded(force: false) }
            return
        }
        notifyDebounceTask?.cancel()
        notifyDebounceTask = Task {
            try? await Task.sleep(nanoseconds: 2_000_000_000)
            await flushIfNeeded(force: false)
        }
    }

    public func flushNow(maxBatches: Int? = nil) async {
        await flushIfNeeded(force: true, maxBatches: maxBatches)
    }

    /// Run one pull cycle (fetch from server, apply to local). No-op if pull or bidirectional store not configured.
    public func pullNow() async {
        guard let pullAPI, let bidir = bidirectionalStore else { return }
        await runPullOnce(pullAPI: pullAPI, store: bidir, mergeStrategy: mergeStrategy)
    }

    private func flushIfNeeded(force: Bool, maxBatches: Int? = nil) async {
        guard !flushing else { return }
        guard !Task.isCancelled else { return }

        if !force, let nextAllowedFlushAt, Date() < nextAllowedFlushAt {
            return
        }
        guard await network.isOnline() else { return }

        do {
            let should = try await shouldFlush(force: force)
            guard should else { return }
        } catch {
            continuation.yield(.uploadFailed(String(describing: error)))
            return
        }

        flushing = true
        defer { flushing = false }

        var batchesProcessed = 0
        while await network.isOnline() {
            if Task.isCancelled { break }
            if let maxBatches, batchesProcessed >= maxBatches { break }

            do {
                let pending = try await store.fetchPendingPoints(limit: batching.maxBatchSize)
                if pending.isEmpty { break }

                let result = try await api.upload(batch: LocationUploadBatch(points: pending))
                let now = Date()

                failureCount = 0
                nextAllowedFlushAt = nil

                if !result.acceptedIds.isEmpty {
                    try await store.markSent(pointIds: result.acceptedIds, sentAt: now)
                }

                if !result.rejected.isEmpty {
                    try await store.markFailed(
                        pointIds: result.rejected.map(\.id),
                        errorMessage: result.rejected.first?.reason ?? "rejected"
                    )
                }

                pendingInsertsSinceLastFlush = 0
                continuation.yield(.didUpload(accepted: result.acceptedIds.count, rejected: result.rejected.count))
                batchesProcessed += 1

                await maybePruneSentPoints(now: now)

                // If server didn't acknowledge any IDs, avoid an infinite loop.
                if result.acceptedIds.isEmpty, result.rejected.isEmpty {
                    break
                }
            } catch {
                scheduleBackoffAfterFailure()
                continuation.yield(.uploadFailed(String(describing: error)))
                break
            }
        }
    }

    private func shouldFlush(force: Bool) async throws -> Bool {
        if force { return true }

        let stats = try await store.pendingStats()
        if stats.count == 0 { return false }
        if stats.count >= batching.maxBatchSize { return true }

        if let oldest = stats.oldestRecordedAt {
            let age = Date().timeIntervalSince(oldest)
            if age >= batching.maxBatchAge { return true }
        }

        return false
    }

    private func scheduleBackoffAfterFailure() {
        failureCount = min(failureCount + 1, 30)

        let baseDelay = retryPolicy.delay(forAttempt: failureCount)
        let jitter = max(0.0, min(1.0, retryPolicy.jitterFraction))
        let factor = 1.0 + Double.random(in: -jitter...jitter)
        let delay = max(0.0, baseDelay * factor)

        nextAllowedFlushAt = Date().addingTimeInterval(delay)
    }

    private func runPullOnce(
        pullAPI: any LocationPullAPI,
        store: any BidirectionalLocationStore,
        mergeStrategy: (any LocationMergeStrategy)?
    ) async {
        do {
            let cursor = try await store.getLastPullCursor()
            let result = try await pullAPI.fetch(since: cursor)
            guard !result.items.isEmpty else {
                if let next = result.nextCursor {
                    try await store.setLastPullCursor(next)
                }
                return
            }
            try await store.applyServerChanges(
                result.items,
                mergeStrategy: mergeStrategy,
                serverTime: result.serverTime,
                lastSyncAt: lastPullAt
            )
            if let next = result.nextCursor {
                try await store.setLastPullCursor(next)
            }
            lastPullAt = Date()
            continuation.yield(.didPull(count: result.items.count))
        } catch {
            continuation.yield(.pullFailed(String(describing: error)))
        }
    }

    private func maybePruneSentPoints(now: Date) async {
        guard let maxAge = retentionPolicy.sentPointsMaxAge else { return }
        guard maxAge > 0 else { return }

        // Don't run a DELETE query on every upload; prune at most once per hour.
        let minInterval: TimeInterval = 60 * 60
        if let lastPruneAt, now.timeIntervalSince(lastPruneAt) < minInterval {
            return
        }

        guard let prunable = store as? any SentPointsPrunableLocationStore else {
            lastPruneAt = now
            return
        }

        do {
            let cutoff = now.addingTimeInterval(-maxAge)
            try await prunable.pruneSentPoints(olderThan: cutoff)
        } catch {
            // Non-fatal: retention cleanup shouldn't block sync.
        }

        lastPruneAt = now
    }
}

File: rtls-kmp/rtls-offline-sync/src/commonMain/kotlin/com/rtls/sync/SyncEngine.kt
package com.rtls.sync

import com.rtls.core.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex

sealed class SyncEngineEvent {
    data class UploadSuccess(val accepted: Int, val rejected: Int) : SyncEngineEvent()
    data class UploadFailed(val message: String) : SyncEngineEvent()
    data class PullSuccess(val count: Int) : SyncEngineEvent()
    data class PullFailed(val message: String) : SyncEngineEvent()
}

class SyncEngine(
    private val store: LocationStore,
    private val api: LocationSyncAPI,
    private val batching: BatchingPolicy,
    private val retryPolicy: SyncRetryPolicy,
    private val retentionPolicy: RetentionPolicy,
    private val networkMonitor: NetworkMonitor?,
    private val scope: CoroutineScope,
    private val pullAPI: LocationPullAPI? = null,
    private val mergeStrategy: LocationMergeStrategy? = null,
    private val pullIntervalSeconds: Long? = null,
    private val pullOnForeground: Boolean = true
) {
    private val _events = MutableSharedFlow<SyncEngineEvent>(extraBufferCapacity = 64)
    val events: SharedFlow<SyncEngineEvent> = _events.asSharedFlow()

    private val bidirectionalStore: BidirectionalLocationStore? = store as? BidirectionalLocationStore

    private var timerJob: Job? = null
    private var networkJob: Job? = null
    private var pullJob: Job? = null
    private var notifyDebounceJob: Job? = null
    private var running = false
    private val flushMutex = Mutex()
    private var failureCount = 0
    private var nextAllowedFlushAtMs: Long? = null
    private var lastPullAtMs: Long? = null
    private var pendingInsertsSinceLastFlush = 0
    private var hasPendingData = false
    private var cachedConnectionType: ConnectionType = ConnectionType.UNKNOWN

    fun start() {
        if (running) return
        running = true

        // Do NOT start timerJob here — it is demand-driven via ensureTimerRunning()

        networkMonitor?.onlineFlow?.let { flow ->
            networkJob = scope.launch {
                flow.collectLatest { online ->
                    cachedConnectionType = networkMonitor.connectionType()
                    if (online && running) {
                        val stats = try { store.pendingStats() } catch (_: Exception) { null }
                        if (stats != null && stats.count > 0) {
                            hasPendingData = true
                            ensureTimerRunning()
                        }
                        flushIfNeeded(force = false, maxBatches = null)
                    }
                }
            }
        }

        if (pullAPI != null && bidirectionalStore != null) {
            val intervalMs = (pullIntervalSeconds ?: 60L).coerceAtLeast(1L) * 1000
            pullJob = scope.launch {
                while (isActive && running) {
                    delay(intervalMs)
                    if (running) runPullOnce()
                }
            }
        }
    }

    fun stop() {
        running = false
        timerJob?.cancel(); timerJob = null
        networkJob?.cancel(); networkJob = null
        pullJob?.cancel(); pullJob = null
    }

    fun notifyNewData() {
        hasPendingData = true
        ensureTimerRunning()

        pendingInsertsSinceLastFlush++
        if (pendingInsertsSinceLastFlush >= batching.maxBatchSize) {
            scope.launch { flushIfNeeded(force = false, maxBatches = null) }
            return
        }

        val debounceMs = if (cachedConnectionType == ConnectionType.CELLULAR) 10_000L else 2_000L
        notifyDebounceJob?.cancel()
        notifyDebounceJob = scope.launch {
            delay(debounceMs)
            flushIfNeeded(force = false, maxBatches = null)
        }
    }

    private fun ensureTimerRunning() {
        if (!hasPendingData || timerJob != null) return
        timerJob = scope.launch {
            while (isActive && running) {
                val baseInterval = batching.flushIntervalSeconds.coerceAtLeast(1L) * 1000
                val interval = if (batching.cellularFlushIntervalSeconds != null &&
                    cachedConnectionType == ConnectionType.CELLULAR) {
                    batching.cellularFlushIntervalSeconds.coerceAtLeast(1L) * 1000
                } else {
                    baseInterval
                }
                delay(interval)
                if (running) flushIfNeeded(force = false, maxBatches = null)
            }
        }
    }

    private fun stopTimerIfDrained() {
        if (hasPendingData) return
        timerJob?.cancel()
        timerJob = null
    }

    suspend fun flushNow(maxBatches: Int? = null) {
        flushIfNeeded(force = true, maxBatches = maxBatches)
    }

    suspend fun pullNow() {
        if (pullAPI == null || bidirectionalStore == null) return
        runPullOnce()
    }

    private suspend fun runPullOnce() {
        val api = pullAPI ?: return
        val bidir = bidirectionalStore ?: return
        try {
            val cursor = bidir.getLastPullCursor()
            val result = api.fetch(cursor)
            if (result.items.isEmpty()) {
                result.nextCursor?.let { bidir.setLastPullCursor(it) }
                return
            }
            bidir.applyServerChanges(result.items, mergeStrategy, result.serverTimeMs, lastPullAtMs)
            result.nextCursor?.let { bidir.setLastPullCursor(it) }
            lastPullAtMs = System.currentTimeMillis()
            _events.emit(SyncEngineEvent.PullSuccess(result.items.size))
        } catch (e: Exception) {
            _events.emit(SyncEngineEvent.PullFailed(e.message ?: "Unknown error"))
        }
    }

    private suspend fun flushIfNeeded(force: Boolean, maxBatches: Int?) {
        if (!running) return
        if (!flushMutex.tryLock()) return
        try {
            val now = System.currentTimeMillis()
            if (!force) {
                nextAllowedFlushAtMs?.let { next -> if (now < next) return }
            }
            networkMonitor?.let { if (!it.isOnline()) return }

            val should = try { shouldFlush(force) } catch (e: Exception) {
                _events.emit(SyncEngineEvent.UploadFailed(e.message ?: "Unknown error"))
                return
            }
            if (!should) return

            var batchesProcessed = 0
            while (running) {
                networkMonitor?.let { if (!it.isOnline()) return }
                if (maxBatches != null && batchesProcessed >= maxBatches) break

                val pending = store.fetchPendingPoints(batching.maxBatchSize)
                if (pending.isEmpty()) {
                    hasPendingData = false
                    stopTimerIfDrained()
                    break
                }

                try {
                    val batch = LocationUploadBatch(points = pending)
                    val result = api.upload(batch)
                    val sentAt = System.currentTimeMillis()
                    failureCount = 0
                    nextAllowedFlushAtMs = null
                    if (result.acceptedIds.isNotEmpty()) {
                        val prunable = store as? SentPointsPrunableLocationStore
                        if (prunable != null) {
                            // Use atomic operation to prevent data loss on crash
                            val olderThanMs = retentionPolicy.sentPointsMaxAgeMs?.let { sentAt - it }
                            prunable.markSentAndPrune(result.acceptedIds, sentAt, olderThanMs)
                        } else {
                            store.markSent(result.acceptedIds, sentAt)
                        }
                    }
                    if (result.rejected.isNotEmpty()) {
                        store.markFailed(result.rejected.map { it.id }, result.rejected.firstOrNull()?.reason ?: "rejected")
                    }
                    pendingInsertsSinceLastFlush = 0
                    _events.emit(SyncEngineEvent.UploadSuccess(result.acceptedIds.size, result.rejected.size))
                    batchesProcessed++
                    if (result.acceptedIds.isEmpty() && result.rejected.isEmpty()) break
                } catch (e: Exception) {
                    scheduleBackoffAfterFailure()
                    _events.emit(SyncEngineEvent.UploadFailed(e.message ?: "Unknown error"))
                    break
                }
            }
        } finally {
            flushMutex.unlock()
        }
    }

    private suspend fun shouldFlush(force: Boolean): Boolean {
        if (force) return true
        val stats = store.pendingStats()
        if (stats.count == 0) {
            hasPendingData = false
            stopTimerIfDrained()
            return false
        }
        if (stats.count >= batching.maxBatchSize) return true
        val oldest = stats.oldestRecordedAtMs ?: return false
        val ageSeconds = (System.currentTimeMillis() - oldest) / 1000
        return ageSeconds >= batching.maxBatchAgeSeconds
    }

    private fun scheduleBackoffAfterFailure() {
        failureCount = (failureCount + 1).coerceAtMost(30)
        val delayMs = retryPolicy.delayForAttempt(failureCount)
        nextAllowedFlushAtMs = System.currentTimeMillis() + delayMs
    }
}
